src/com/amazon/ion/impl/lite/IonContainerLite.java

```diff
@@ -109,8 +73,124 @@ public IonSystemLite getSystem() {
     @Override
     public abstract void accept(ValueVisitor visitor) throws Exception;
 
+    // Context to be used when cloning a value tree by walking it iteratively.
+    private static class CloneContext {
+
+        // The parent of the value currently being cloned.
+        IonContainerLite parentOriginal = null;
+
+        // The copy of the parent value, into which values at the current depth are placed.
+        IonContainerLite parentCopy = null;
+
+        // True if the parent value is a struct. Storing this in a variable is a performance optimization.
+        boolean parentIsStruct = false;
+
+        // The copied context relevant to the values cloned at the current depth.
+        IonContext contextCopy = null;
+
+        // The index in the parent container of the value currently being cloned.
+        int childIndex = 0;
+    }
+
+
+    // Doubles the size of a stack of CloneContext objects.
+    private static CloneContext[] growStack(CloneContext[] stack) {
+        CloneContext[] newStack = new CloneContext[stack.length * 2];
+        System.arraycopy(stack, 0, newStack, 0, stack.length);
+        return newStack;
+    }
+
+    /**
+     * Prepares a CloneContext for the container that is about to be walked.
+     * @param cloneContext the context to fill in.
+     * @param containerOriginal the container being cloned.
+     * @param containerCopy the container clone.
+     * @param isDatagram true if the container being cloned is a datagram.
+     */
+    private static void setCloneContext(CloneContext cloneContext, IonContainerLite containerOriginal, IonContainerLite containerCopy, boolean isDatagram) {
+        cloneContext.parentOriginal = containerOriginal;
+        cloneContext.parentCopy = containerCopy;
+        containerCopy._children = new IonValueLite[containerOriginal._children.length];
+        containerCopy._child_count = containerOriginal._child_count;
+        if (isDatagram) {
+            cloneContext.contextCopy = TopLevelContext.wrap(containerOriginal._context.getContextSymbolTable(), (IonDatagramLite) containerCopy);
+            cloneContext.parentIsStruct = false;
+        } else {
+            cloneContext.contextCopy = containerCopy;
+            cloneContext.parentIsStruct = containerCopy instanceof IonStructLite;
+        }
+        cloneContext.childIndex = 0;
+    }
+
+    /**
+     * Clones the container by walking it iteratively.
+     * @return the clone.
+     */
+    final IonContainerLite deepClone() {
```

Comment:
I found that passing in a boolean to deepClone when the value being cloned is a datagram is performance-neutral and allows us to avoid the `instanceof` check at the top. However, passing in the entire initial context was 1-5% slower, so I added a comment about that and went with the boolean.

Regarding factoring, this is the cleanest one I've come up with given how much method-internal state there is. Splitting it up into smaller methods tends to require helpers with many parameters that run into issues with single-return. This factoring basically follows the template that we used for the other recursive-to-iterative overhauls. That said, if you know of a better way please propose it.

I've added some additional comments within deepClone to hopefully guide readers more clearly.